### PR TITLE
HWKAPM-485 Use a typed exception for triggering retries

### PR DESCRIPTION
**Pull request by objectiser**
Review notes:
- Added typed exception for both the processor, to trigger the retry, but also on the TraceService and AnalyticsService APIs related to failures when storing information.

### CODE REVIEW

**Comment by objectiser** [comment]
@metlos Hi Lukas, would you be able to review this one?

File: server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsCacheMDB.java [hunk 1]
```diff
@@ -67,7 +68,8 @@ public void init() {
         setProcessor(new AbstractProcessor<CommunicationDetails,Void>(ProcessorType.ManyToMany) {
 
             @Override
-            public List<Void> processManyToMany(String tenantId, List<CommunicationDetails> items) throws Exception {
+            public List<Void> processManyToMany(String tenantId, List<CommunicationDetails> items)
+                    throws RetryAttemptException {
                 communicationDetailsCache.store(tenantId, items);
```
Comment (metlos) [inline server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsCacheMDB.java]: Is the lack of a try block here because the`StoreException` is never thrown by `CommunicationsDetailsCache` or have you forgotten to refactor this method?

File: api/src/main/java/org/hawkular/apm/api/services/AnalyticsService.java [hunk 2]
```diff
@@ -190,50 +190,51 @@
...
 
     /**
+     * This method returns the list of host names where activities were executed, subject to the supplied
+     * criteria.
+     *
+     * @param tenantId The tenant id
+     * @param criteria The criteria
+     * @return The list of host names
+     */
+    List<String> getHostNames(String tenantId, Criteria criteria);
```
Comment (metlos) [inline api/src/main/java/org/hawkular/apm/api/services/AnalyticsService.java]: This seems unrelated to the "theme" of this PR. I'm much worse than this when it comes to keeping focus of the PRs, so I actually don't mind :wink: Just pointing out it is discouraged.

File: api/src/main/java/org/hawkular/apm/api/services/StoreException.java [hunk 3]
```diff
@@ -14,25 +14,44 @@
...
+package org.hawkular.apm.api.services;
 
 /**
- * This class is an exception representing Elasticsearch failures.
+ * This exception indicates a failure to store information.
  *
  * @author gbrown
  */
-public class ElasticsearchFailures extends Exception {
+public class StoreException extends Exception {
```
Comment (metlos) [inline api/src/main/java/org/hawkular/apm/api/services/StoreException.java]: Nitpick: Given this is a checked exception and is used in IO-related scenarios, maybe this could be replaced by the `IOException` to save 1 class definition. On the other hand the name is maybe more descriptive.

**Comment by metlos** [comment]
LGTM apart from the small suggestion at the `StoreException`.

File: server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsCacheMDB.java (hunk 1 above)
Comment (objectiser) [inline server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsCacheMDB.java]: Correct - however I think the current cache API needs to define a checked exception - I'll create a separate jira for this.

File: api/src/main/java/org/hawkular/apm/api/services/AnalyticsService.java (hunk 2 above)
Comment (objectiser) [inline api/src/main/java/org/hawkular/apm/api/services/AnalyticsService.java]: I know, thought I would sneak it in :)

**Comment by objectiser** [comment]
@metlos Thanks!
